v2/async_runner.go

```diff
@@ -0,0 +1,74 @@
+package bugsnag
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"sync"
+)
+
+type AsyncRunner[T any] struct {
+	ch        chan T
+	wg        sync.WaitGroup
+	processFn ProcessFunc[T]
+}
+
+type ProcessFunc[T any] func(T)
+
+var ErrChannelFull = errors.New("channel is full")
+
+func NewAsyncRunner[T any](
+	ctx context.Context,
+	channelSize int,
+	workerCount int,
+	processFn ProcessFunc[T],
+) (*AsyncRunner[T], error) {
+	if channelSize < 0 {
+		return nil, fmt.Errorf("channelSize must be non negative")
+	}
+
+	b := &AsyncRunner[T]{
+		ch:        make(chan T, channelSize),
+		wg:        sync.WaitGroup{},
+		processFn: processFn,
+	}
+
+	b.startWorkers(ctx, workerCount)
+	return b, nil
+}
+
+func (b *AsyncRunner[T]) Submit(v T) error {
```

Comment:
An alternative design would be to submit the function to be executed, instead of the message to be processed. Let me know which one you prefer.
```
func (b *AsyncRunner) Submit(fn func()) error
```

Reply:
Another idea is to replace AsyncRunner with the library https://github.com/alitto/pond, which achieves the same goal.